### PR TITLE
#12: Ruby on Rails ORM example app

### PR DESCRIPTION
**Pull request by wasif-shehzad**
https://github.com/YugaByte/orm-examples/issues/12

This PR adds Ruby on Rails ORM example app.

### CODE REVIEW

File: ruby/ror/public/robots.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+# See http://www.robotstxt.org/robotstxt.html for documentation on how to use the robots.txt file
```
Comment (syedshah12tech) [inline ruby/ror/public/robots.txt]: remove this file

**Comment by wasif-shehzad** [comment]
> I see alot of .keep files. Do we need them?

> .keep files only track the directory existence for git. Extra directories are removed now

**Comment by syedshah12tech** [comment]
LGTM

File: ruby/ror/.gitignore [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#
+# If you find yourself ignoring temporary files generated by your text editor
+# or operating system, you probably want to add a global ignore instead:
+#   git config --global core.excludesfile '~/.gitignore_global'
+
+# Ignore bundler config.
+/.bundle
+
+# Ignore all logfiles and tempfiles.
+/log/*
```
Comment (rkarthik007) [inline ruby/ror/.gitignore]: Do we need these? Are they absolute paths?

File: ruby/ror/.gitignore [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+# Ignore uploaded files in development
+/storage/*
+!/storage/.keep
+
+.byebug_history
+
+# Ignore master key for decrypting credentials and more.
+/config/master.key
+../.idea/
```
Comment (rkarthik007) [inline ruby/ror/.gitignore]: Do we need these directories outside the pwd?

File: ruby/ror/Gemfile [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+gem 'puma', '~> 3.11'
+# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
+# gem 'jbuilder', '~> 2.5'
+# Use Redis adapter to run Action Cable in production
+# gem 'redis', '~> 4.0'
+# Use ActiveModel has_secure_password
+# gem 'bcrypt', '~> 3.1.7'
+
+# Use ActiveStorage variant
+# gem 'mini_magick', '~> 4.8'
```
Comment (rkarthik007) [inline ruby/ror/Gemfile]: Lets remove commented out lines

File: ruby/ror/app/controllers/application_controller.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+class ApplicationController < ActionController::API
+end
```
Comment (rkarthik007) [inline ruby/ror/app/controllers/application_controller.rb]: Lets remove the concerns folder if its not needed

File: ruby/ror/app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
+    class OrdersController < ApplicationController
+
+     def index
+        orders = Order.all
+       render json:{content:orders},status: :ok
+        end
+
+
+        def create
+                req_products = params[:products]
```
Comment (rkarthik007) [inline ruby/ror/app/controllers/orders_controller.rb]: Indentation is off on various lines, lets use 2 spaces everywhere.

File: ruby/ror/app/controllers/orders_controller.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                user_id = params[:userId]
+              productIDs = req_products.map{|p|  p[:productId]; }
+              orderInfo = [];
+
+                db_products = Product.where(id:productIDs);
+                total=0;
+               db_products.each{|pro| orderedProduct = req_products.select{|prod| pro[:id] == prod[:productId]}
```
Comment (rkarthik007) [inline ruby/ror/app/controllers/orders_controller.rb]: If over 120 chars, lets split into 2 lines

Comment (rkarthik007) [inline ruby/ror/app/controllers/orders_controller.rb]: Indent this as a loop block

File: ruby/ror/bin/setup [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+chdir APP_ROOT do
+  # This script is a starting point to setup your application.
+  # Add necessary setup steps to this file.
+
+  puts '== Installing dependencies =='
+  system! 'gem install bundler --conservative'
+  system('bundle check') || system!('bundle install')
+
+  # puts "\n== Copying sample files =="
```
Comment (rkarthik007) [inline ruby/ror/bin/setup]: Lets remove commented out lines

File: ruby/ror/bin/spring [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env ruby
```
Comment (rkarthik007) [inline ruby/ror/bin/spring]: Do we need this file?

File: ruby/ror/.gitignore (hunk 2 above)
Comment (syedshah12tech) [reply ruby/ror/.gitignore]: we checked, a leading slash in gitignore doesn't mean absolute path. 

As stated in the documentation: https://git-scm.com/docs/gitignore
"A leading slash matches the beginning of the pathname. For example, "/*.c" matches "cat-file.c" but not "mozilla-sha1/sha1.c"."

**Comment by rkarthik007** [comment]
@syedshah12tech - could you please comment when this diff is ready to review? @ttyusupov said he will take a look.

**Comment by syedshah12tech** [comment]
@rkarthik007 yes, I will let you know as soon as its ready for your review

**Comment by syedshah12tech** [comment]
thanks @wasif-shehzad @rkarthik007 the PR is now ready for your review

**Comment by wasif-shehzad** [comment]
Removed dependencies on pgcrypto and uuid-ossp. Changed products.id and users.id columns to product.productId and users.userId respectively

**Comment by ttyusupov** [comment]
Why do we need empty file `ruby/ror/tmp/.keep` under git?

File: ruby/ror/config/database.yml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+
+default: &default
+  adapter: postgresql
+  encoding: unicode
+  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+
+development:
+  <<: *default
+  database: postgres
+  username: postgres
```
Comment (ttyusupov) [inline ruby/ror/config/database.yml]: @rkarthik007 do we want to update database/username to something other?

Comment (rkarthik007) [reply ruby/ror/config/database.yml]: Thought about it, lets sweep through all the different orm's and normalize it afterwards. For now, this is ok.

File: ruby/ror/Gemfile [hunk 11]
```diff
@@ -0,0 +1,19 @@
+source 'https://rubygems.org'
+git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+ruby '2.4.1'
```
[the remark this answers is not in the document]
Comment (syedshah12tech) [reply ruby/ror/Gemfile]: Good point. We don't really require any specific ruby version. So will remove this line

File: ruby/ror/Gemfile [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+gem 'rails', '~> 5.2.3'
+gem 'pg', '>= 0.18', '< 2.0'
```
[the remark this answers is not in the document]
Comment (syedshah12tech) [reply ruby/ror/Gemfile]: It seems libpq is not installed on your host. Do you have Postgresql installed?

[the remark this answers is not in the document]
Comment (ttyusupov) [reply ruby/ror/Gemfile]: Fixed it by running first:
```
gem install pg -- --with-pg-config=<yugabyte-install-dir>/postgres/bin/pg_config
```

File: ruby/ror/Gemfile [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
+end
+
+group :development do
+  gem 'listen', '>= 3.0.5', '< 3.2'
+  gem 'spring'
+  gem 'spring-watcher-listen', '~> 2.0.0'
+end
+
+gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
```
[the remark this answers is not in the document]
Comment (ttyusupov) [reply ruby/ror/Gemfile]: Can/should we do something about it?

[the remark this answers is not in the document]
Comment (syedshah12tech) [reply ruby/ror/Gemfile]: missed this comment earlier. Will get back soon

[the remark this answers is not in the document]
Comment (syedshah12tech) [reply ruby/ror/Gemfile]: warning fixed

[the remark this answers is not in the document]
Comment (ttyusupov) [reply ruby/ror/Gemfile]: Do we actually need these platforms for Ruby on Rails?

[the remark this answers is not in the document]
Comment (syedshah12tech) [reply ruby/ror/Gemfile]: Yes why not? Ruby supports those platforms, we should too.

**Pull request merged**